llms/mlx_lm/utils.py

```diff
@@ -143,17 +143,21 @@ def sample(logits: mx.array) -> Tuple[mx.array, float]:
         if temp == 0:
             token = mx.argmax(logits, axis=-1)
         else:
+            # referenced implementation from https://github.com/huggingface/transformers/blob/main/src/transformers/generation/logits_process.py#L449-L460
             if top_p > 0 and top_p < 1.0:
                 if (
                     logits.dtype == mx.bfloat16
                 ):  # workdaround for unable to load kernel contiguous_scan_inclusive_sum_bfloat16_bfloat16
                     logits = logits.astype(mx.float32)
+
                 probs = mx.softmax(logits / temp, axis=-1)
 
-                sorted_probs = mx.sort(probs)[::-1]
-                sorted_indices = mx.argsort(probs)[::-1]
+                # sort probs in ascending order
+                sorted_probs = mx.sort(probs, axis=-1)
+                sorted_indices = mx.argsort(probs, axis=-1)
                 cumulative_probs = mx.cumsum(sorted_probs, axis=-1)
```

Comment:
I still think you need to flip the probabilities here. You want the smallest set of tokens with total mass > p.

Reply:
> I still think you need to flip the probabilities here. You want the smallest set of tokens with total mass > p.

@awni would you please elaborate on this? Do you mean that when using cumsum, you want the cumulative sum to be in descending order?

Reply:
Cool, it seems there are no major blockers. I will start working on adding the test.